bot/utils/function.py

```diff
@@ -73,3 +75,43 @@ def get_bound_args(func: t.Callable, args: t.Tuple, kwargs: t.Dict[str, t.Any])
     bound_args.apply_defaults()
 
     return bound_args.arguments
+
+
+def update_wrapper_globals(wrapper: types.FunctionType, wrapped: types.FunctionType) -> types.FunctionType:
+    """
+    Update globals of `wrapper` with the globals from `wrapped`.
+
+    For forwardrefs in command annotations discordpy uses the __global__ attribute of the function
+    to resolve their values, with decorators that replace the function this breaks because they have
+    their own globals.
+
+    This function creates a new function functionally identical to `wrapper`, which has the globals replaced with
+    a merge of `wrapped`s globals and the `wrapper`s globals.
+
+    In case a global name from `wrapped` conflicts with a name from `wrapper`'s globals, `wrapper` will win
+    to keep it functional, but this may cause problems if the name is used as an annotation and
```

Comment:
I'm looking at this now, it should be possible by checking the intersection of wrapper's co_names and the annotations.
What do you think about raising an error that can optionally be turned into a warning if there's a conflict? I'm not sure how discoverable it'd be in tests for the devs when developing a command and it completely breaks the parameter that uses the conflicting global

Reply:
Detecting it during automated testing in CI is more ideal than getting an error during runtime. A test would have to get all defined commands (can at least filter by ones that have this decorator) and do some magic to detect the conflict for each one.

Reply:
It will cause somewhat cryptic errors at runtime if the command is used so I'm not sure if I agree here.

Reply:
You think a cryptic runtime error is better than a clear error during CI which would prevent the buggy code from even making it into production?

Reply:
Is there a good way of checking what the functions are decorated with without parsing their source or setting some attribute on the objects?

Reply:
The current error is cryptic, which would be solved by a better runtime error/warning before it's reached. The test will prevent buggy code getting into production but when developing the tests have to be ran manually which probably is not the first course of action after getting a traceback while developing something new.
If discord.py fails to resolve the annotation in some way the error stops the bot's execution and looks like this https://paste.fuelrats.com/dumemageri.tex which doesn't tell all that much without some research

Reply:
I don't think so. After all, the whole point of wrapping it is to make it indistinguishable. Anyway, I shouldn't impose this on you now. If you can't find a solution now, it can be postponed for a future PR. If you can generate a good runtime warning, then that's better than nothing for now.

Reply:
Addressed it in e86e9f921a4bbbe42a5fb6fd8486425f11af62cf, there may be some edge case false positives to the error because of what `co_names` stores but the error should clear up what's going on if a conflict occurs.